// rc-swipeout/es/Swipeout.js
import _extends from 'babel-runtime/helpers/extends';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
var __rest = this && this.__rest || function (s, e) {
    var t = {};
    for (var p in s) {
        if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
    }if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
        if (e.indexOf(p[i]) < 0) t[p[i]] = s[p[i]];
    }return t;
};
import React from 'react';
import ReactDOM from 'react-dom';
import Hammer from 'rc-hammerjs';
import omit from 'omit.js';

var Swipeout = function (_React$Component) {
    _inherits(Swipeout, _React$Component);

    function Swipeout(props) {
        _classCallCheck(this, Swipeout);

        var _this = _possibleConstructorReturn(this, (Swipeout.__proto__ || Object.getPrototypeOf(Swipeout)).call(this, props));

        _this.onPanStart = _this.onPanStart.bind(_this);
        _this.onPan = _this.onPan.bind(_this);
        _this.onPanEnd = _this.onPanEnd.bind(_this);
        _this.onCloseSwipe = _this.onCloseSwipe.bind(_this);
        _this.openedLeft = false;
        _this.openedRight = false;
        return _this;
    }

    _createClass(Swipeout, [{
        key: 'componentDidMount',
        value: function componentDidMount() {
            this.btnsLeftWidth = this.left ? this.left.offsetWidth : 0;
            this.btnsRightWidth = this.right ? this.right.offsetWidth : 0;
            document.body.addEventListener('touchstart', this.onCloseSwipe, true);
        }
    }, {
        key: 'componentWillUnmount',
        value: function componentWillUnmount() {
            document.body.removeEventListener('touchstart', this.onCloseSwipe, true);
        }
    }, {
        key: 'onCloseSwipe',
        value: function onCloseSwipe(ev) {
            var _this2 = this;

            if (this.openedLeft || this.openedRight) {
                var pNode = function (node) {
                    while (node.parentNode && node.parentNode !== document.body) {
                        if (node.className.indexOf(_this2.props.prefixCls + '-actions') > -1) {
                            return node;
                        }
                        node = node.parentNode;
                    }
                }(ev.target);
                if (!pNode) {
                    ev.preventDefault();
                    this.close();
                }
            }
        }
    }, {
        key: 'onPanStart',
        value: function onPanStart(e) {
            this.panStartX = e.deltaX;
            this.panStartY = e.deltaY;
        }
    }, {
        key: 'onPan',
        value: function onPan(e) {
            var posX = e.deltaX - this.panStartX;
            var posY = e.deltaY - this.panStartY;
            if (Math.abs(posX) <= Math.abs(posY)) {
                return;
            }
            var _props = this.props,
                left = _props.left,
                right = _props.right;

            if (posX < 0 && right.length) {
                this.swiping = true;
                this._setStyle(Math.min(posX, 0));
            } else if (posX > 0 && left.length) {
                this.swiping = true;
                this._setStyle(Math.max(posX, 0));
            }
        }
    }, {
        key: 'onPanEnd',
        value: function onPanEnd(e) {
            if (!this.swiping) {
                return;
            }
            this.swiping = false;
            var _props2 = this.props,
                _props2$left = _props2.left,
                left = _props2$left === undefined ? [] : _props2$left,
                _props2$right = _props2.right,
                right = _props2$right === undefined ? [] : _props2$right;

            var btnsLeftWidth = this.btnsLeftWidth;
            var btnsRightWidth = this.btnsRightWidth;
            var posX = e.deltaX - this.panStartX;
            var openLeft = posX > btnsLeftWidth / 2;
            var openRight = posX < -btnsRightWidth / 2;
            if (openRight && posX < 0 && right.length) {
                this.open(-btnsRightWidth, false, true);
            } else if (openLeft && posX > 0 && left.length) {
                this.open(btnsLeftWidth, true, false);
            } else {
                this.close();
            }
        }
        // left & right button click

    }, {
        key: 'onBtnClick',
        value: function onBtnClick(ev, btn) {
            var onPress = btn.onPress;
            if (onPress) {
                onPress(ev);
            }
            if (this.props.autoClose) {
                this.close();
            }
        }
    }, {
        key: '_getContentEasing',
        value: function _getContentEasing(value, limit) {
            // limit content style left when value > actions width
            if (value < 0 && value < limit) {
                return limit - Math.pow(limit - value, 0.85);
            } else if (value > 0 && value > limit) {
                return limit + Math.pow(value - limit, 0.85);
            }
            return value;
        }
        // set content & actions style

    }, {
        key: '_setStyle',
        value: function _setStyle(value) {
            var limit = value > 0 ? this.btnsLeftWidth : -this.btnsRightWidth;
            var contentLeft = this._getContentEasing(value, limit);
            this.content.style.left = contentLeft + 'px';
            if (this.cover) {
                this.cover.style.display = Math.abs(value) > 0 ? 'block' : 'none';
                this.cover.style.left = contentLeft + 'px';
            }
        }
    }, {
        key: 'open',
        value: function open(value, openedLeft, openedRight) {
            if (!this.openedLeft && !this.openedRight && this.props.onOpen) {
                this.props.onOpen();
            }
            this.openedLeft = openedLeft;
            this.openedRight = openedRight;
            this._setStyle(value);
        }
    }, {
        key: 'close',
        value: function close() {
            if ((this.openedLeft || this.openedRight) && this.props.onClose) {
                this.props.onClose();
            }
            this._setStyle(0);
            this.openedLeft = false;
            this.openedRight = false;
        }
    }, {
        key: 'renderButtons',
        value: function renderButtons(buttons, _ref) {
            var _this3 = this;

            var prefixCls = this.props.prefixCls;
            return buttons && buttons.length ? React.createElement(
                'div',
                { className: prefixCls + '-actions ' + prefixCls + '-actions-' + _ref, ref: function ref(el) {
                        return _this3[_ref] = el;
                    } },
                buttons.map(function (btn, i) {
                    return React.createElement(
                        'div',
                        { key: i, className: prefixCls + '-btn ' + (btn.hasOwnProperty('className') ? btn.className : ''), style: btn.style, role: 'button', onClick: function onClick(e) {
                                return _this3.onBtnClick(e, btn);
                            } },
                        React.createElement(
                            'div',
                            { className: prefixCls + '-btn-text' },
                            btn.text || 'Click'
                        )
                    );
                })
            ) : null;
        }
    }, {
        key: 'render',
        value: function render() {
            var _this4 = this;

            var _a = this.props,
                prefixCls = _a.prefixCls,
                left = _a.left,
                right = _a.right,
                disabled = _a.disabled,
                children = _a.children,
                restProps = __rest(_a, ["prefixCls", "left", "right", "disabled", "children"]);
            var divProps = omit(restProps, ['autoClose', 'onOpen', 'onClose']);
            var refProps = {
                ref: function ref(el) {
                    return _this4.content = ReactDOM.findDOMNode(el);
                }
            };
            return (left.length || right.length) && !disabled ? React.createElement(
                'div',
                _extends({ className: '' + prefixCls }, divProps),
                React.createElement('div', { className: prefixCls + '-cover', ref: function ref(el) {
                        return _this4.cover = el;
                    } }),
                this.renderButtons(left, 'left'),
                this.renderButtons(right, 'right'),
                React.createElement(
                    Hammer,
                    _extends({ direction: 'DIRECTION_HORIZONTAL', onPanStart: this.onPanStart, onPan: this.onPan, onPanEnd: this.onPanEnd }, refProps),
                    React.createElement(
                        'div',
                        { className: prefixCls + '-content' },
                        children
                    )
                )
            ) : React.createElement(
                'div',
                _extends({}, refProps, divProps),
                children
            );
        }
    }]);

    return Swipeout;
}(React.Component);

Swipeout.defaultProps = {
    prefixCls: 'rc-swipeout',
    autoClose: false,
    disabled: false,
    left: [],
    right: [],
    onOpen: function onOpen() {},
    onClose: function onClose() {}
};
export default Swipeout;
